src/analysis/processing/qgsalgorithmexportmeshvertices.cpp

```diff
@@ -0,0 +1,301 @@
+  mTransform = QgsCoordinateTransform( meshLayer->crs(), outputCrs, context.transformContext() );
+  if ( !meshLayer->nativeMesh() )
+    meshLayer->updateTriangularMesh( mTransform ); //necessary to load the native mesh
+
+  mNativeMesh = *meshLayer->nativeMesh();
+
+  QVariant datasetGroupIndexVariant = parameters[QStringLiteral( "DATASET_GROUPS" )];
+
+  if ( datasetGroupIndexVariant.type() != QVariant::List )
+    return false;
+
+  QVariantList datasetGroupsList = datasetGroupIndexVariant.toList();
+
+  if ( feedback )
+  {
+    if ( feedback->isCanceled() )
+      return false;
+    feedback->setProgress( 0 );
+    feedback->setProgressText( QObject::tr( "Preparing data" ) );
+  }
+
+  QDateTime layerReferenceTime = static_cast<QgsMeshLayerTemporalProperties *>( meshLayer->temporalProperties() )->referenceTime();
+
+  //! Extract the date time used to export dataset values under a relative time
+  QgsInterval relativeTime;
+  QVariant parameterTimeVariant = parameters.value( QStringLiteral( "DATASET_TIME" ) );
+  if ( !parameterTimeVariant.isValid() || parameterTimeVariant.type() != QVariant::Map )
+    return false;
+
+  QVariantMap parameterTimeMap = parameterTimeVariant.toMap();
+  if ( !parameterTimeMap.contains( QStringLiteral( "type" ) ) )
+    return false;
+
+  if ( parameterTimeMap.value( QStringLiteral( "type" ) ) == QStringLiteral( "static" ) )
+  {
+    relativeTime = 0;
+  }
+  else if ( parameterTimeMap.value( QStringLiteral( "type" ) ) == QStringLiteral( "dataset-time-step" ) )
+  {
+    QVariant datasetIndexVariant = parameterTimeMap.value( QStringLiteral( "value" ) );
+    if ( datasetGroupIndexVariant.type() != QVariant::List )
+      return false;
+    QVariantList datasetIndexVariantSplit = datasetIndexVariant.toList();
+    if ( datasetIndexVariantSplit.count() != 2 )
+      return false;
+    QgsMeshDatasetIndex datasetIndex( datasetIndexVariantSplit.at( 0 ).toInt(), datasetIndexVariantSplit.at( 1 ).toInt() );
+    relativeTime = meshLayer->datasetRelativeTime( datasetIndex );
+  }
+  else
+  {
+    QVariant dateTimeVariant = parameterTimeMap.value( QStringLiteral( "value" ) );
+    if ( dateTimeVariant.type() != QVariant::DateTime )
+      return false;
+    QDateTime dateTime = dateTimeVariant.toDateTime();
+    relativeTime = QgsInterval( layerReferenceTime.secsTo( dateTime ) );
+  }
+
+  for ( int i = 0; i < datasetGroupsList.count(); ++i )
+  {
+    int  groupIndex = datasetGroupsList.at( i ).toInt();
+    QgsMeshDatasetIndex datasetIndex = meshLayer->datasetIndexAtRelativeTime( relativeTime, groupIndex );
+
+    DataGroup dataGroup;
+    dataGroup.metadata = meshLayer->datasetGroupMetadata( datasetIndex );
+    int dataCount = dataGroup.metadata.dataType() == QgsMeshDatasetGroupMetadata::DataOnVertices ? mNativeMesh.vertexCount() : mNativeMesh.faceCount();
+    dataGroup.datasetValues = meshLayer->datasetValues( datasetIndex, 0, dataCount );
```

Comment:
It could be quite expensive with very big meshes, but it is quite reasonable  comparing with other operation. As in the processAlgorithm we have to be thread safe, I think we don't have a lot of choice, we prepare here the data that are thread safe.
I think other solutions will ask deeper modification in the data provider to make it thread safe.

Reply:
Could you potentially clone the mesh layer in prepare, and then use the clone in the process step?

Reply:
The problem is to bring the data provider, especially MDAL provider...

Reply:
Is there any "in memory" mesh layer concept? If not, then couldn't you re-load a new mesh layer using the same uri?

Reply:
Actually, now cloning the mesh is possible, and the data provider is recreated from the same uri. But, for now it is not a real cloning because there are some data that are not cloned as extra persistent dataset loaded in the provider and other extra dataset like virtual dataset. I was thinking about make a real cloning by writing a dom element with writeXml and then readXml with the new mesh. It will need some refactorization of the write/read xml but seems possible.

But for now I am not completly convince of this way :
We could have a new mesh layer that could be used only in the process with a thread safe way, but this new layer needs :
- recreate a new data provider, actually MDAL needs to load the mesh on its side, and with a lot of driver, it needs to load all the dataset groups too.
- to reload completly a new native mesh in QGIS side (btw make change in the mesh layer implementation to allow reloading mesh without make the triangular mesh that consume ressources)
- if simplification is enable, restart a mesh simplification that could consume a lot of ressources. But here, maybe we could avoid this simplification  for the new mesh.
Even all that could be done in the process thread, it will consume quite much more memory, especially for big meshes.

With the actual implementation, in the algorithm preparation, what is the more expensive is the transfer of the data from the data provider side to QGIS side, and just for the selected dataset. Afer all, it is "just" a copy of data without calculation, and it is not so long for ordinary mesh (it is what is done for the renderer in the main thread).

So cloning the mesh layer could reduce the preparation to almost nothing but will need more resources (time and memory) on the process thread and that will need also more or less complex modification in the mesh layer implementation.

With the actual implementation of the algorithm, it spends as much time in the main thread as the renderer and it consume just the memory needed to store the needed dataset.

Reply:
Much clearer with tracing :

- a coarse mesh (2.9 millions of faces) with only one dataset group exported:
![image](https://user-images.githubusercontent.com/7416892/97812183-04e7d700-1c56-11eb-8aad-702bc1d66bd2.png)
Preparing (little pink one) : 18.6 ms.

- a less big mesh (730 000 faces), but with a lot of dataset group exported (20 groups):
![image](https://user-images.githubusercontent.com/7416892/97812249-7758b700-1c56-11eb-8ee6-58d89bb98fc4.png)
Preparing (little pink one) : 1.1 s.

20 dataset groups is quite big, and I think it is a extreme case

Reply:
Ok, let's give it a try...